working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+It is an error if an optional parameter (named or otherwise) with no default
+value has a potentially non-nullable type.
+
+It is an error if a required named parameter has a default value.
+
+It is an error if a named parameter that is part of a `required` group is not
+bound to an argument at a call site.
+
+It is an error to call the default `List` constructor with a length argument and
+a type argument which is potentially non-nullable.
+
+For the purposes of errors and warnings, the null aware operators `?.` and `?..`
+are checked as if the receiver of the operator had non-nullable type.
+
+It is an error for a class to extend, implement, or mixin a type of the form
+`T?` for any `T`.
+
+It is an error for a class to extend, implement, or mixin the type `Never`.
+
+It is an error to call a method, setter, or getter on a receiver of static type
+`Never` (including via a null aware operator).
+
+It is an error to apply an expression of `Never` in the function position in a
+function call.
+
+It is an error if the static type of `e` in the expression `throw e` is
+potentially nullable.
+
+It is not an error for the body of a `lazy` field to reference `this`.
+
+It is an error for a toplevel variable, a static field, or a formal parameter to
+be declared `lazy`.
+
+It is an error if the type `T` in the **on-catch** clause `on T catch` is
+potentially nullable.
+
+### Assignability
+
+The definition of assignability is changed as follows.  
+
+A type `T` is **assignable** to a type `S` if `T` is `dynamic`, or if `S` is a
+subtype of `T`.
+
+### Generics
+
+The default bound of generic type parameters is treated as `Object?`.
+
+### Type promotion, Definite Assignment, and Definite Completion
+
+**TODO** Fill this out.
+
+### Null promotion
+
+The machinery of type promotion is extended to promote the type of variables
+based on nullability checks subject to the same set of restrictions as normal
+promotion.  The relevant checks and the types they are considered to promote to
+are as follows.
+
+A check of the form `e == null` or of the form `e is Null` where `e` has static
+type `T` promotes the type of `e` to `Null` in the `true` continuation, and to
+**NonNull**(`T`) in the
+`false` continuation.
+
+A check of the form `e != null` or of the form `e is T` where `e` has static
+type `T?` promotes the type of `e` to `T` in the `true` continuation, and to
+`Null` in the `false` continuation.
+
+The static type of an expression `e!` is **NonNull**(`T`) where `T` is the
+static type of `e`.
+
+The **NonNull** function defines the null-promoted version of a type, and is
+defined as follows.
+
+- **NonNull**(Null) = Never
+- **NonNull**(_C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>) = _C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>  for class *C* other than Null
+- **NonNull**(FutureOr<_T_>) = FutureOr<_T_>   
+- **NonNull**(_T_<sub>0</sub> Function(...)) = _T_<sub>0</sub> Function(...)
+- **NonNull**(Function) = Function
+- **NonNull**(Object) = Object
```

Comment:
Technically `Object` is handled by the _C_ case above.

Reply:
Ok - called it out there though to make it clear.  I prefer technically clear to technically correct.  ;)